### ISSUE TITLE
Several reassembly error cases on Ramblr

### ISSUE THREAD

**Issue by witbring** [angr/angr#3549]
When I tested Ramblr (commit 64d1049, Apr. 2022) for my research, I found several interesting bug cases.
 
First, we observed that Ramblr makes wrong symbolic expressions so some recompiled binaries refer to incorrect addresses. As an example, given the disassembly code 'movb $8, 0x820824b' found in as-new of binutils (x86), Ramblr symbolized the immediate 0x820824b as 'label_1807+3'. However, 'label_1807' was defined at the address 0x820824b so the symbolic expression points to the address 0x820824e (=0x820824b+3).  
- binary file: [as-new_x86_o0.tar.gz](https://github.com/angr/patcherex/files/9689904/as-new_x86_o0.tar.gz)

```
    # 0x8058455:    movb    $8, 0x820824b [IMM, MEM]
    movb    $8, (label_1807+3)

    # data @ 0x820824b
label_1807:
    .long .label_27340-4
```

Second, we found that Ramblr emits duplicate code. For example, Ramblr generates a function 'sub_80a0400' twice when it reassembles as-new of binutils (x86). As a result, the bug causes a compilation error.
- biniary file: [as-new_x86_o2.tar.gz](https://github.com/angr/patcherex/files/9689903/as-new_x86_o2.tar.gz)

```
sub_80a0400:
    pushl   %ebp
    # 0x80a0401:    pushl   %edi [REG]
    pushl   %edi
    # ...

sub_80a0400:
    pushl   %ebp
    # 0x80a0401:    pushl   %edi [REG]
    pushl   %edi
    # ...
```

Third, we observed that Ramblr generates multiple definitions of single labels. As an example, Ramblr generates labels 'label_197' and 'label_199' multiple times while reassembling ptx in coreutils (x86). The bug also causes a compilation error. 
- binary file: [ptx.tar.gz](https://github.com/angr/patcherex/files/9689888/ptx.tar.gz)

```
    # 0x804a8ac:    cmpb    $0, 0x805402e(%eax) [IMM, MEM]
    cmpb    $0, (label_197+2)(%eax)
    # ...

    # 0x804a8ba:    cmpb    $0, 0x805402f(%eax) [IMM, MEM]
    cmpb    $0, (label_199+3)(%eax)
    # ...

    # data @ 0x805402c
label_195:
label_196:
label_197:
label_199:
    .long .label_1223
    # data @ 0x805402e
label_197:
    .byte 4
    # data @ 0x805402f
label_199:
    .byte 8

    # data @ 0x8054030
```

Lastly, we found that Ramblr omits some definitions of labels. For example, given the data pointer 0x555946 found in as-new of binutils (x64), Ramblr symbolized the pointer as 'label_32525' instead of 'label_32501'. Moreover, Ramblr misses the definition of the label 'label_32525' so it causes a compilation error. 
- binary file: [as-new_x64_o2.tar.gz](https://github.com/angr/patcherex/files/9689901/as-new_x64_o2.tar.gz)

```
    # data @ 0x555946
label_32501:
    # ...

    # data @ 0x555ff0
    .quad label_32525    
```
This is a data dump of the target binary.
```
$ objdump -s -j .rodata binutils-2.31.1/x64/gcc/nopie/o0-gold/bin/as-new | grep 555ff0 
 555ff0 48595500 00000000 48595500 00000000  HYU.....HYU.....
```

**Comment by ltfish** [comment]
I'm transferring this issue to angr/angr since Reassembler is solely implemented as an angr analysis.

I believe all issues except for the first one are bugs in the current implementation of Reassembler. While I do not have cycles to debug or fix them, I do intend to fix them at some point (if no one else gets to them before I do).

**Comment by witbring** [comment]
Thank you for your reply.
I hope it will help to improve Ramblr.

**Comment by ChenLijiong** [comment]
Hello, as for the Ramblr framework source code and use method you mentioned above, can you share it？

**Comment by witbring** [comment]
Hi @ChenLijiong. 
We've made our tool, Reassessor, available on [GitHub](https://github.com/SoftSec-KAIST/Reassessor) for public access. Also, [our paper](https://www.usenix.org/conference/usenixsecurity23/presentation/kim-hyungseok) provides a comprehensive overview of Reassessor's design. I trust this will be beneficial for your work.

**Comment by ChenLijiong** [comment]
Thank you very much. I will try to use seassesor. After a brief look at the introduction of this, I need a reorganized program before using it. According to what you mentioned, you used the framework ramblr to reorganize the program, so is ramblr an open source framework? If so, where do I get it and use it

**Comment by witbring** [comment]
Yes, it is. You'd better to refer to [the sample code](https://github.com/angr/patcherex/issues/39#issuecomment-1018726303).
